### PR TITLE
Remove superfluous (???) files from sdist

### PR DESCRIPTION
**Pull request by samuelcolvin**
Will this, a bunch of files are removed from the sdist (`.tar.gz`) file which will be uploaded to PyPI.

## Question

* Is this correct?
* should tests still be included?
* I'm excluding `/docs/`, I think this is better, is there a strong argument against this?

With this changes, the following files are included in `pydantic-2.0.0.dev0.tar.gz`:
```
pydantic-2.0.0.dev0
├── .gitignore
├── HISTORY.md
├── LICENSE
├── Makefile
├── PKG-INFO
├── pydantic
│  ├── __init__.py
│  ├── _internal
│  │  ├── __init__.py
│  │  ├── ... 3 more files
│  ├── ... 23 more files
├── pyproject.toml
├── README.md
├── requirements
│  ├── 10 files
└── tests
   ├── mypy
   │  ├── 34 files
   ├── pyright
   │  ├── ... 2 files
   ├── ... 34 files
```

### CODE REVIEW

**Comment by samuelcolvin** [comment]
@ofek with your knowledge of packaging, does this look correct to you?

@davidbrochart any idea if this will work okay with conda-forge?

please review.

File: pyproject.toml [hunk 1]
```diff
@@ -5,6 +5,16 @@ build-backend = 'hatchling.build'
...
 path = 'pydantic/version.py'
 
+[tool.hatch.build]
+exclude = [
+    '/.github/**',
+    '/changes/**',
+    '/docs/**',
+    '/mkdocs.yml',
+    '/.pre-commit-config.yaml',
+    '/setup.py',
```
Comment (samuelcolvin) [inline pyproject.toml:15]: excluding `.gitignore` seems to be ignored :shrug:.

**Comment by davidbrochart** [comment]
> @davidbrochart any idea if this will work okay with conda-forge?

Yes, we will just need to add `hatchling` to the host requirements.

**Comment by samuelcolvin** [comment]
great, thanks.

File: pyproject.toml [hunk 2]
```diff
@@ -5,6 +5,16 @@ build-backend = 'hatchling.build'
 [tool.hatch.version]
 path = 'pydantic/version.py'
 
+[tool.hatch.build]
+exclude = [
+    '/.github/**',
+    '/changes/**',
+    '/docs/**',
```
Comment (ofek) [inline pyproject.toml:12]:
```suggestion
    '/.github',
    '/changes',
    '/docs',
```

**Comment by samuelcolvin** [comment]
@mgorny (and anyone else who's involved in distributing pydantic downstream - linux distros etc.) is this okay with you?

Do you / would you, use the `docs/` directory?

**Comment by mgorny** [comment]
We're not using sdists, we're fetching autogenerated archives from GitHub directly, so it shouldn't affect us.

**Comment by samuelcolvin** [comment]
Interesting, thanks.

**Comment by mgorny** [comment]
No problem, and thanks for asking. It's probably not something specific to pydantic, we're using GitHub archives whenever possible because sdists are missing tests too often. At this point, not even trying them saves us time. Not to mention that pypi URLs are just horrible.

**Comment by samuelcolvin** [comment]
So (although gentoo doesn't use sdist), you would say we're right to include tests in the sdist and right (or at least it's okay) to omit docs?

**Comment by mgorny** [comment]
If you asked me a year ago, I'd say that sdist should include both tests and docs (in source form) but I have non-Python background and I look at sdist as something akin to autotools' `make dist`, i.e. something you should be able to use to build, test and install everything's that useful about the package. As packagers, we find tests useful, and I know that many Gentoo users do appreciate offline documentation.

That said, I know that many people using Python see sdists as source material for `pip install` only, and prefer that they are as small as possible.

In other words, I see the points of both sides here and I don't really have a 100% good answer here. You can always try to satisfy everyone and e.g. push minimal sdist to pypi, and "fuller" sdist to GitHub ;-).

**Comment by ofek** [comment]
> you would say we're right to include tests in the sdist and right (or at least it's okay) to omit docs?

That's what I do

File: pyproject.toml [hunk 3]
```diff
@@ -5,6 +5,16 @@ build-backend = 'hatchling.build'
 [tool.hatch.version]
 path = 'pydantic/version.py'
 
+[tool.hatch.build]
+exclude = [
```
Comment (ofek) [inline pyproject.toml:9]: maybe instead:

```toml
include = [
    '/pydantic',
    '/tests',
    '/Makefile',
]
```

Comment (ofek) [inline pyproject.toml:9]: maybe instead:

```toml
[tool.hatch.build.targets.sdist]
include = [
    '/pydantic',
    '/tests',
    '/Makefile',
]
```

https://hatch.pypa.io/latest/plugins/builder/sdist/#default-file-selection

Comment (samuelcolvin) [reply pyproject.toml]: done.

**Comment by samuelcolvin** [comment]
thanks everyone, merging this.
